### PR TITLE
[core] Add task and object reconstruction status to ray memory

### LINKED ISSUE #21427: [Core] [Feature] Better observability for lineage reconstruction
### Search before asking

- [X] I had searched in the [issues](https://github.com/ray-project/ray/issues) and found no similar feature requirement.

### Description

Right now, you can only see how many objects have been recovered in the past. This is not very useful when trying to debug correctness or performance issues, such as if object reconstruction is slow/hanging.

Some useful things to collect are:
- ObjectRefs that are currently being reconstructed (not including any intermediate ObjectRefs that had to be recursively reconstructed)
- the current frontier of ObjectRefs that are being reconstructed
- how long each object reconstruction is taking, maybe compared to the initial execution

It would also be nice to integrate an ObjectRef's current status (e.g., re-executing, pending dependency reconstruction, etc) with `ray memory`/the dashboard.

### Use case

Debugging correctness or performance issues in object reconstruction.

### Related issues

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by stephanie-wang**
## Why are these changes needed?

Improve observability for general objects and lineage reconstruction by adding a "Status" field to `ray memory`. The value of the field can be:
```
  // The task is waiting for its dependencies to be created.
  WAITING_FOR_DEPENDENCIES = 1;
  // All dependencies have been created and the task is scheduled to execute.
  SCHEDULED = 2;
  // The task finished successfully.
  FINISHED = 3;
```

In addition, tasks that failed or that needed to be re-executed due to lineage reconstruction will have a field listing the attempt number. Example output:
```
IP Address    | PID      | Type    | Call Site | Status    | Size     | Reference Type | Object Ref
192.168.4.22  | 279475   | Driver  | (task call) ... | Attempt #2: FINISHED | 10000254.0 B | LOCAL_REFERENCE | c2668a65bda616c1ffffffffffffffffffffffff0100000001000000


```

## Related issue number
Closes #21427.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by stephanie-wang** [comment]
I'd also like to add some metrics like total number of reconstructing tasks to the CoreWorkerStats or debug_state.txt, but I'm not really sure how to read those metrics.

File: src/ray/protobuf/common.proto [hunk 1]
```diff
@@ -477,6 +477,25 @@ message ViewData {
...
+  // The task is waiting for its dependencies to be created.
+  WAITING_FOR_DEPENDENCIES = 1;
+  // All dependencies have been created and the task is scheduled to execute.
+  SCHEDULED = 2;
+  // The task finished successfully.
+  FINISHED = 3;
+  // The task finished previously but is now being scheduled again because its
+  // outputs need to be reconstructed. It is waiting for its dependencies to be
+  // created again.
+  RECONSTRUCTING_AND_WAITING_FOR_DEPENDENCIES = 4;
```
Comment (ericl) [inline src/ray/protobuf/common.proto:493]: Should there be a "reconstructing" flag instead of separate states?

**Review by ericl — COMMENTED** [review]
Does test_memstat.py also need updating?

File: src/ray/protobuf/common.proto (hunk 1 above)
Comment (stephanie-wang) [reply src/ray/protobuf/common.proto:493]: I wanted to avoid adding an extra field to the protobuf but I can do this if you think it's cleaner.

**Comment by stephanie-wang** [comment]
cc @rkooo567 

By the way, I'm going to hold off on adding this to any metrics for now because I'm not really sure if it'll make sense. These stats are collected at the owners so they don't have anything to do with the tasks that are queued at the local raylet.

**Review by ericl — APPROVED** [review]

**Comment by ericl** [comment]
I like the attempt number change!

File: src/ray/core_worker/task_manager.h [hunk 2]
```diff
@@ -47,6 +48,8 @@ class TaskFinisherInterface {
       const std::vector<ObjectID> &inlined_dependency_ids,
       const std::vector<ObjectID> &contained_ids) = 0;
 
+  virtual void MarkDependenciesResolved(const TaskID &task_id) = 0;
```
Comment (rkooo567) [inline src/ray/core_worker/task_manager.h:51]: Personal preference, but why don't we use `OnDependenciesResolved`? Mark sounds like we will have some operations for the task id

**Review by rkooo567 — COMMENTED** [review]
Can you update the public documentation (ray memory) to explain the meaning states/attempt?

**Pull request merged**